Custom Fields option: Add confirmation step

marekhrabe opened the pull request:
Fixes #15320.

"Custom Fields" in Options modal requires a full page reload in order to toggle the option. Currently, this happens instantly after clicking the checkbox. 

This PR changes the flow to add an inline confirmation step:

see the latest version https://github.com/WordPress/gutenberg/pull/15688#issuecomment-515697517

marekhrabe commented:
Update:

This PR shows additional message about a reload and a confirmation button to proceed. This additional UI only shows when a reload action is required and disappears when you change your mind and toggle the option back without a confirmation. It is additionally announced with spoken messages to differentiate the tickbox from all others which save options instantly.

It's not well styled and polished at this point. Let's figure out a proper interaction first.

Question: Do spoken messages make sense here? If so, what would their formulations be? I'm sure I haven't came up with the most ideal copy.

If we confirm this approach is a way forward, I can take a look at styling, making string translatable etc…

noisysocks commented:
I can't answer the question about using `speak()`, but functionally I think this is the right approach!

I added the `Needs Design` label so that the design team can offer some suggestions. Personally I'd advocate for something relatively inline and unobtrusive.

<img width="383" alt="Screen Shot 2019-05-27 at 11 22 06" src="https://user-images.githubusercontent.com/612155/58389924-c2d66100-8071-11e9-92ae-751f5af2dfed.png">

(The link would be a `<Button>` with `isLink`.)

mapk commented:
Oh dang, turns out I added a redesign to the issue and not the PR. In case people don't want to click around, here's the proposed redesign.

![Screen Shot 2019-05-28 at 6 58 44 PM](https://user-images.githubusercontent.com/617986/58563047-46619080-81df-11e9-9a64-360166646807.png)

I'm not entirely sure if the notification should be `blue` or `yellow`. Any thoughts on that?

mapk commented:
Okay, because this is more of a warning notification, let's go with a yellow version.

![Screen Shot 2019-05-29 at 10 07 20 AM](https://user-images.githubusercontent.com/617986/58576839-0eb41200-81fa-11e9-93a9-b2b687e91cd8.png)

afercia commented:
Discussed during today's accessibility bug scrub. Quick feedback: if there’s text with relevant information and if the button says “Save and Reload” there’s no need for a spoken message.

noisysocks commented:
![confirmation](https://user-images.githubusercontent.com/612155/60694789-a5bf6880-9f22-11e9-9760-88360d908b61.gif)

I've rebased this, removed the `speak()` commands, and changed the UI to look like the mockup posted above.

It was a little tricky getting everything to line up nicely. I ended up using CSS grid for this. @mapk, let me know what you think and feel free to make any adjustments to the styling.

noisysocks commented on packages/components/src/checkbox-control/index.js:
```diff
@@ -27,6 +37,7 @@ function CheckboxControl( { label, className, heading, checked, help, instanceId
 			<label className="components-checkbox-control__label" htmlFor={ id }>
 				{ label }
 			</label>
+			{ children }
```
Not sure about this change. None of the other control components have `children` as a prop.

noisysocks committed:
Add confirmation step to Custom Fields option

Add a confirmation step to the Custom Fields option so that the page
doesn't reload without warning.

Co-authored-by: Marek Hrabe <marekhrabe@me.com>

noisysocks commented:
![confirmation](https://user-images.githubusercontent.com/612155/60694789-a5bf6880-9f22-11e9-9760-88360d908b61.gif)

I've rebased this, removed the `speak()` commands, and changed the UI to look like the mockup posted above. It should be ready for a review from both a designer and developer 🙂

noisysocks committed:
Update E2E tests for new Custom Fields setting flow

karmatosed commented:
Can I ask why we're using an information message there and not just having as a simple extra text section? It feels a lot having even the yellow notice on that settings page. 

For example, in accessibility options this is being considered, to me, this could fit better.


![](https://user-images.githubusercontent.com/253067/60364323-bf6e2680-99dd-11e9-922f-34966bdb47d2.png)

afercia committed:
Remove notice in favor of plain text.

afercia commented:
Removed the notice style in favor of plain text. Adjusted margins. Screenshot:

<img width="792" alt="Screenshot 2019-07-08 at 14 30 23" src="https://user-images.githubusercontent.com/1682452/60811576-e243cc00-a18f-11e9-9758-bf81b8a36512.png">

Note: when pressed, the button gets disabled with a HTML `disabled` attribute. This an exception to the pattern used in Gutenberg to avoid focus losses, as in this case there's a page reload anyways.

mapk commented:
Just to jump back into this and provide devs with a more high fidelity mockup, I put one together based on both @karmatosed's and @afercia's comments. Thanks for the feedback!

I think it's still important to keep the margin for readability, but dropping the notification style makes a lot of sense here.

![Screen Shot 2019-07-10 at 9 50 05 AM](https://user-images.githubusercontent.com/617986/60988395-61541400-a2f8-11e9-89d1-47f1fdfe6eeb.png)

noisysocks committed:
Custom Fields option: Adjust margin to align text

noisysocks commented:
Done and done!

![custom-fields](https://user-images.githubusercontent.com/612155/61014438-c1b38600-a3ca-11e9-9273-da2b70faf922.gif)

This should be ready for a final code review now.

talldan commented on packages/edit-post/src/components/options-modal/options/enable-custom-fields.js:
```diff
@@ -1,47 +1,54 @@
+		<BaseOption
+			label={ label }
+			isChecked={ isChecked }
+			onChange={ ( nextIsChecked ) => setIsChecked( nextIsChecked ) }
```
could this be simplified to `onChange={ setIsChecked }`?

talldan commented on packages/edit-post/src/components/options-modal/options/enable-custom-fields.js:
```diff
@@ -1,47 +1,54 @@
+					document.getElementById( 'toggle-custom-fields-form' ).submit();
+				} }
+			>
+				{ __( 'Save & Reload' ) }
```
'Save & Reload' - clicking the button does save the change to the custom fields option, but doesn't save any other edits to the post, which may be misleading to the user and lead to lost data. I'm thinking that perhaps the post should be saved when this is clicked?

marekhrabe replied:
In case there is unsaved content, the page won't reload automatically and will still show the "Are you sure" dialog of the browser. While this is not ideal, I think that merging this will still improve the totally unclear and inaccessible way that is shipped now. I think that mixing together post saving could become very complex (you need to wait for the request to succeed etc) and we would need to handle possible edge cases. 

Do you think that updating the label to remove the word "Save" would solve the problem in a way we can get this merged and fix the root accessibility problem? I'm thinking of making the label dynamic based on the checkbox status to reflect the exact action that will be taken — either "Enable & Reload" or "Disable & Reload".

marekhrabe replied:
And/or add something in lines of "Make sure your editor content is saved before proceeding"

talldan commented on packages/edit-post/src/components/options-modal/options/enable-custom-fields.js:
```diff
@@ -1,47 +1,54 @@
+	return (
+		<>
+			<p className="edit-post-options-modal__custom-fields-confirmation-message">
+				{ __( 'Page reload is required for this change.' ) }
```
Very very minor, but I think this minor change makes the sentence sound a lot less terse.
```suggestion
				{ __( 'A page reload is required for this change.' ) }
```

talldan reviewed (commented):
I had one big question about 'Save & Reload', so haven't approved this, but generally the code looks good. 

I know that previously clicking the custom fields option didn't save the post, but I think the messaging now will lead to some expecting that to happen.

karmatosed commented:
Can I get confirmation that this does 'save and reload'? I am reading comments and wondering if that is the case @noisysocks and @talldan. If it doesn't actually do that saying so is a little unexpected.

mapk commented:
> I know that previously clicking the custom fields option didn't save the post, but I think the messaging now will lead to some expecting that to happen.

I believe the "save" refers to the setting. It's saving the setting and reloading the page. But it does cause confusion when the help text says that "A *page* reload is required" and then the button implies a "Save and load" as if referring to the _page_.

Maybe we can adjust the language of the button? Here are some ideas:

`Submit`
`Agree`
`Acknowledge`
`Reload`
`Do it`

Any other thoughts?

talldan commented:
Agree that adjusting the wording would be a good compromise to get this merged. I like @marekhrabe's suggestions in the comment above, or just shortening it to 'Reload' / 'Reload now'. 

Could it also be worth adding something to the main part of the message? e.g. 'A page reload is required for this change, please save any changes to your content before reloading.'

marekhrabe committed:
update description

Co-Authored-By: Daniel Richards <daniel.richards@automattic.com>

marekhrabe committed:
simplify change handler

marekhrabe committed:
add dynamic button based on the next state to be saved

mapk commented:
> "Enable & Reload" or "Disable & Reload"

Oh, I missed these. I prefer them over my suggestions. Thanks!

> Could it also be worth adding something to the main part of the message? e.g. 'A page reload is required for this change, please save any changes to your content before reloading.'

If handling the save of the document automatically before reloading gets tricky, this seems like the right solution. And if they miss that text and trigger the reload anyways, alerting them with an "Are you sure" message as @marekhrabe suggested works too.

marekhrabe committed:
add notice about saved content and make sure it won't expand modal

marekhrabe commented:
> "Enable & Reload" or "Disable & Reload"

I have implemented these in 8320d22c83a3d51ae5d14500a002698173e84aaf

> And if they miss that text and trigger the reload anyways, alerting them with an "Are you sure" message as @marekhrabe suggested works too.

"Are you sure" functionality is done on a global level and it is already triggering for this option. 

I have added an extra sentence to the info message about saved content.

# Current State

<img width="365" alt="Screenshot 2019-07-27 at 09 37 12" src="https://user-images.githubusercontent.com/156676/61997161-7f3cab00-b052-11e9-94c8-9be67bbf53f6.png">

<img width="368" alt="Screenshot 2019-07-27 at 09 37 01" src="https://user-images.githubusercontent.com/156676/61997163-82379b80-b052-11e9-8009-5f14b9e7e66f.png">

With unsaved content:

<img width="433" alt="Screenshot 2019-07-27 at 09 37 31" src="https://user-images.githubusercontent.com/156676/61997164-8794e600-b052-11e9-8537-dfc613fc9a5c.png">

marekhrabe committed:
update snapshot with new button label and confirm msg

marekhrabe committed:
make button label in test dynamic

marekhrabe commented:
All tests now account for new labels. Are we good to go with this state https://github.com/WordPress/gutenberg/pull/15688#issuecomment-515697517?

noisysocks commented on packages/edit-post/src/components/options-modal/options/enable-custom-fields.js:
```diff
@@ -1,47 +1,54 @@
+					document.getElementById( 'toggle-custom-fields-form' ).submit();
+				} }
+			>
+				{ nextState === true ? __( 'Enable & Reload' ) : __( 'Disable & Reload' ) }
```
This could be simplified.

```suggestion
				{ nextState ? __( 'Enable & Reload' ) : __( 'Disable & Reload' ) }
```

noisysocks commented on packages/edit-post/src/components/options-modal/options/enable-custom-fields.js:
```diff
@@ -1,47 +1,54 @@
-	render() {
-		const { label } = this.props;
-		const { isChecked } = this.state;
+export function CustomFieldsConfirmation( { nextState } ) {
```
Am thinking something like `willEnable` is a more descriptive name for this prop, but it's not a big deal since this component is only used in this file.

marekhrabe replied:
It took me the longest time to come up with the bad name. `willEnable` is 💯

noisysocks reviewed (approved):
The code changes look good, and this tests well. Thanks for the assist @marekhrabe!

I'll leave it for @mapk to give the final say on whether the design and copy are ready to be merged.

marekhrabe committed:
simplify condition for determining label

Co-Authored-By: Robert Anderson <robert@noisysocks.com>

marekhrabe committed:
use willEnable as the prop name

mapk commented:
I gave it another go. The added sentence looks great! Thanks everyone for helping out with this one. :shipit:

mapk reviewed (approved):
:shipit: Looks good!

marekhrabe commented:
This PR has truly been a journey 😀Thanks everybody!

marekhrabe commented:
I just saw this featured as one of the tips about Gutenberg on Twitter 😀 

https://twitter.com/editorskit/status/1178693225923497984
